source/_posts/2019-02-18-building-a-decentralized-reddit-with-embark-part-3.md

```diff
@@ -0,0 +1,627 @@
+
+
+<pre class="javascript"><code>import { List } from './List';
+
+export class App extends Component {
+
+  render() {
+    return (
+      &lt;React.Fragment&gt;
+        &lt;h1&gt;DReddit&lt;/h1&gt;
+        &lt;CreatePost /&gt;
+        &lt;List /&gt;
+      &lt;/React.Fragment&gt;
+    )
+  }
+}
+</code></pre>
+
+### Fetching posts data
+
+Let's fill our new `List` component with life! As mentioned earlier, we'll use our Smart Contract's `numPosts()` method to get hold of the total number of posts available. We then use that number to iterate over all indices and request every post individually. Since this is logic we want to execute once `List` component is ready, we'll use its `componentDidMount()` method for that:
+
+<pre class="javascript"><code>export class List extends Component {
+  ...
+  async componentDidMount() {
+    const totalPosts = await DReddit.methods.numPosts().call();
+
+    let list = [];
+
+    if (totalPosts > 0) {
+      for (let i = 0; i < totalPosts; i++) {
+        const post = DReddit.methods.posts(i).call();
+        list.push(post);
+      }
+    }
+
+    list = await Promise.all(list);
+  }
+  ...
+}</code></pre>
+
+Notice that in the above code we don't `await` the calls to every individual post. This is on purpose as we don't want to wait on each and every promise to resolve, but first collect all of the promises we need and then resolve them all in one go using `Promise.all().`
+
+Last but not least, we need to add an `id` property to every post as mentioned earlier. This is easily done by simply iterating over all posts and assigning the post's index as `id`. Once that is done, we can use `setState()` to update our component's state and render the list:
+
+<pre class="javascript"><code>async componentDidMount() {
+  ...
+  list = list.map((post, index) => {
+    post.id = index;
+    return post;
+  });
+
+  this.setState({ posts: list });
+}</code></pre>
+
+That's it! Our application now renders a list of all created posts. Unfortunately, posts are not being re-fetched  automatically when adding new posts. For the time being, we'll have to reload the browser every time after adding a post. However, this we'll address now.
+
+### Reloading posts
+
+There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's not communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.
```

Comment:
```suggestion
There is certainly different ways to make the list of posts update automatically, so take the following approach with a grain of salt. What we need is a way to have the `createPost` component tell the `List` component to reload its posts. However, there's no communication layer in place when building a simple React app like this, so the most straight forward way to make this possible, is to move the logic of loading the posts in the parent component of `CreatePost` and `List` (in our case `App`), and have it pass that logic down to places where its needed. This also means we'll be fetching the list inside `App` and pass down the pure data to `List`.
```

Reply:
Also, I just thought, there is another way to do it. You can always use contract `events` in the List and listen to new posts.

Reply:
Yea good point!